patmat/src/main/scala/patmat/Huffman.scala

```diff
@@ -105,15 +130,24 @@ trait Huffman extends HuffmanInterface:
    * In such an invocation, `until` should call the two functions until the list of
    * code trees contains only one single tree, and then return that singleton list.
    */
-  def until(done: List[CodeTree] => Boolean, merge: List[CodeTree] => List[CodeTree])(trees: List[CodeTree]): List[CodeTree] = ???
+  def until(done: List[CodeTree] => Boolean, merge: List[CodeTree] => List[CodeTree])(trees: List[CodeTree]): List[CodeTree] = {
+    if(done(trees)){
+      trees
+    }else{
+      until(done, merge)(merge(trees))
+    }
+  }
 
   /**
    * This function creates a code tree which is optimal to encode the text `chars`.
    *
    * The parameter `chars` is an arbitrary text. This function extracts the character
    * frequencies from that text and creates a code tree based on them.
    */
-  def createCodeTree(chars: List[Char]): CodeTree = ???
+  def createCodeTree(chars: List[Char]): CodeTree = {
+    val trees: List[CodeTree] = makeOrderedLeafList(times(chars))
+    until(singleton, combine)(trees).head
```

Comment:
```suggestion
    until(singleton, combine)(trees).head // I didn't follow the logic, but is there a possibility that the answer will be an empty list and then you `head` it?
```

Reply:
> 
The answer will not be empty list since the "done" criteria is set to "singleton" which ensures the result is of size 1.
It is however possible the initial input was empty, and yes, I should've checked that